### PR TITLE
Enable leaving FetchingOrder dialog

### PR DESCRIPTION
**Pull request by malinajirka**
Parent issue #3726

This PR allows the user to leave "Fetching order dialog" during the payment flow. This dialog is shown only when the user leaves PaymentSuccessful screen before the order gets refetched. 
In other words, if the payment succeeds
- User presses back and refetching order finished - exit the payment flow
- User presses back and refetching order has not finished - show fetching order dialog
- User presses back when fetching order dialog is shown - exit the payment flow and show "data might be outdate" message

To test:
1. Open a detail of an order in US dollars
2. Go through the payment flow
3. Use charles to suspend re-fetch order request (or add a delay here)
4. Press back to dismiss PaymentSuccessful screen
5. Notice a progress dialog is shown
6. Press back again
7. Notice the payment flow is dismissed and a snackbar message is shown

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt [hunk 1]
```diff
@@ -247,8 +249,17 @@ class CardReaderPaymentViewModel @Inject constructor(
...
-        return if (fetchOrderJob?.isActive == true) {
-            viewState.value = FetchingOrderState
+        if (fetchOrderJob?.isActive == true) {
+            if (viewState.value != FetchingOrderState) {
+                viewState.value = FetchingOrderState
+            } else {
+                // show "data might be outdated" and exit the flow when the user presses back on FetchingOrder screen
+                launch {
+                    triggerEvent(ShowSnackbar(R.string.card_reader_fetching_order_failed))
+                    delay(1) // needs to be here so both events are consumed
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt]: Usually, it's a bad idea to have any kind of delay in production code. There is no guarantee that it actually will always work. 

I would just create an event `ShowSnackbarAndExit` just for sake of avoiding this.

Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt]: Also, you are on the main thread here, and `triggerEvent` uses `setValue`. Are you sure there is no guarantee of the order of delivery of events in this case? (that's a lot of `of` in this sentence =) )

**Review by kidinov — COMMENTED** [review]
Looks good! I would just reconsider this delay approach

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt (hunk 1 above)
Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt]: > Are you sure there is no guarantee of the order of delivery of events in this case? (that's a lot of of in this sentence =) )

I'll double check this.

>  There is no guarantee that it actually will always work.

I agree, but the worst case scenario, the snackbar won't be shown.

> I would just create an event ShowSnackbarAndExit just for sake of avoiding this.

I'd personally like to avoid this - this might lead to way more issues (aka poor design decisions). If we are not ok that there is a very small chance the snackbar might not get shown, we might want to consider launching a new coroutine. Wdyt?

Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt]: > I agree, but the worst case scenario, the snackbar won't be shown.

Actually, I take that back - I don't think it can ever go wrong. This delay is not there to wait some time, it's there just to make sure the task is put at the end of the queue. We could wait 0.000000000000000001 and it'd still work as expected. The only case when it wouldn't work is if the runtime completely ignored the delay, which I don't think can happen. Wdyt?
Having said that, I agree that the code doesn't look great, so if we could come up with a cleaner approach it'd definitely be good.

Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt]: > Are you sure there is no guarantee of the order of delivery of events in this case? (that's a lot of of in this sentence =) )

I'll double check this, but if it won't work, I'll try using `yeild`. That should be equivalent of using `delay(1)` but a bit more readable/cleaner.

Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt]:
```
Actually, I take that back - I don't think it can ever go wrong. This delay is not there to wait some time, it's there just to make sure the task is put at the end of the queue. We could wait 0.000000000000000001 and it'd still work as expected.
```

Are you talking about the looper message queue? I'm not sure that I'm following you here. From what I see `LiveData.setValue` synchronously notifies active observers. Basically, I don't see why you need a coroutine here. Just

```
 } else {
                // show "data might be outdated" and exit the flow when the user presses back on FetchingOrder screen
                    triggerEvent(ShowSnackbar(R.string.card_reader_fetching_order_failed))
                    triggerEvent(Exit)
```

should be working just fine

Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt]: > Are you talking about the looper message queue? I'm not sure that I'm following you here. From what I see LiveData.setValue synchronously notifies active observers. 

I wanted to double check that it actually works in our codebase. The specific part I wanted to double check was [MultiLiveEvent](https://github.com/woocommerce/woocommerce-android/blob/3b3a7eaaa38d7b99c398c732aae1b37a62a89f15/WooCommerce/src/main/kotlin/com/woocommerce/android/viewmodel/MultiLiveEvent.kt#L29) and its `pending` field. I was afraid the second event wouldn't get delivered because `pending == false`. It's not the case, so I agree we are safe to remove `delay`. I had a bit different problem with `MultiLiveEvent` a few weeks ago and I mixed it with this issue.

So to summarize -> after checking MultiLiveEvent I can confirm that the delay is not necessary at all. Great catch! ;)

Fixed in e2711874842ab42f7283affe4f18b24ca0518646

**Review by kidinov — APPROVED** [review]
Looks good!
